src/search.c

```diff
@@ -0,0 +1,108 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <assert.h>
+#include <stddef.h>
+#include <simclist.h>
+#include "search.h"
+#include "match.h"
+
+int find_match(char* line, char* word, int pos_start, int lineNum, list_t* matches) {
+    if (strlen(word) > strlen(line)) return -1;
+
+    char* dup = strdup(line);
+
+    char* line2 = strdup(line);
+    int pos = pos_start;
+    char* token = strtok(line2, " ,.!?\r\t\n");
+    match* foundMatch = NULL;
+
+    while (token != NULL) {
+
+        if (strncmp(token, word, strlen(word)) == 0) {
+
+            /* Config match */
+            foundMatch = match_new(token, lineNum, pos, dup);
+            match_append(foundMatch, matches);
+
+            /* Clean-up */
+            free(dup);
+
+            /* Return */
+            return pos;
+        }
+
+        pos += strlen(token) + 1;
+        token = strtok(NULL, " ,.!?\r\t\n");
+
+    }
+
+    return -1;
+
+}
+
+list_t* parse_file_buffered(FILE* pf, int start_line, int end_line, char* word, list_t* matches) {
+
+    char* line = NULL;
+    int wordlen = strlen(word);
+    size_t len = 0;
+    ssize_t read;
+
+    int found = -1;
+    int lineNum = start_line;
+
+    while (lineNum <= end_line && (read = getline(&line, &len, pf)) != -1) {
+
+        char sanitized[strlen(line) + 1];
+        strncpy(sanitized, line, strlen(line));
+        sanitized[strcspn(sanitized, "\r\n")] = 0;
+        char* dupLine = strdup(sanitized);
+
+        char line2[strlen(sanitized) + 1];
+        strncpy(line2, sanitized, strlen(sanitized));
+        found = find_match(sanitized, word, 1, lineNum, matches);
+
+        while (found != -1 && found + wordlen < read) {
+            memset(sanitized, ' ', found + wordlen);
+            found = find_match(sanitized, word, found + wordlen + 2, lineNum, matches);
+
+            if (strncmp(sanitized, dupLine, strlen(sanitized)) != 0) {
```

Comment:
You can't use `strlen` inside of a safe string function. Safe string functions bound at a certain length to avoid overflowing in the case of a missing null terminator. `strlen` works by counting up until you hit a null terminator. If there's no null terminator, `strlen` either segfaults or returns garbage, which is what we're trying to avoid with `strncmp`. In general, safe string functions should always reference a compile-time constant (`#define`).